Source/Core/Common/FileUtil.cpp

```diff
@@ -516,27 +533,81 @@ bool DeleteDirRecursively(const std::string& directory)
   return success;
 }
 
-// Create directory and copy contents (optionally overwrites existing files)
-bool CopyDir(const std::string& source_path, const std::string& dest_path, const bool destructive)
+bool Copy(std::string_view source_path, std::string_view dest_path, bool overwrite_existing)
 {
+  DEBUG_LOG_FMT(COMMON, "{}: {} --> {} ({})", __func__, source_path, dest_path,
+                overwrite_existing ? "overwrite" : "preserve");
+
   auto src_path = StringToPath(source_path);
   auto dst_path = StringToPath(dest_path);
-  if (fs::equivalent(src_path, dst_path))
+  std::error_code error;
+  auto options = fs::copy_options::recursive;
+  if (overwrite_existing)
+    options |= fs::copy_options::overwrite_existing;
+  fs::copy(src_path, dst_path, options, error);
+  if (error)
+  {
+    std::error_code error_ignored;
+    if (fs::equivalent(src_path, dst_path, error_ignored))
+      return true;
+
+    ERROR_LOG_FMT(COMMON, "{}: failed {} --> {} ({}): {}", __func__, source_path, dest_path,
+                  overwrite_existing ? "overwrite" : "preserve", error.message());
+    return false;
+  }
+  return true;
+}
+
+static bool Move(const std::filesystem::path& src, const std::filesystem::path& dst,
+                 std::error_code& error)
+{
+  fs::rename(src, dst, error);
+  if (!error)
     return true;
 
-  DEBUG_LOG_FMT(COMMON, "{}: {} --> {}", __func__, source_path, dest_path);
+  // rename failed, try fallbacks
```

Comment:
Like I said, I'm not sure if the `copy_file` part is useful, but the directory iteration can definitely happen if you try to merge into an existing directory.

Reply:
Like, to be clear: You have `/d1/a.txt` and `/d2/b.txt`. You then try to `fs::rename("/d1", "/d2")`. This fails; `/d2` is non-empty so the exception does not apply. But the expected behavior of a move from `/d1` to `/d2` is that all contents from `/d1` end up in `/d2` -- you want to end up with a `/d2` containing both `a.txt` and `b.txt` -- so you have to iterate through the contents and move them one-by-one using the same algorithm.

Reply:
Sure.

Reply:
Wait, I just read the documentation on this again... I see why you were confused now, because it says for the folder -> folder case `If not an error, then new_p is first deleted`. That's not what we want here, the merge is the desired behavior, so I guess this needs an extra check for that. Weird that this apparently differs between platforms, on Windows it behaves like I described.

Reply:
Hm... well whatever the case, it's written in a very confusing way.

I'll try this on a Linux system in a bit just to be sure.

Reply:
Okay yeah your reading is correct, if the destination is a non-empty directory it fails on Linux too.

```
#include <filesystem>
#include <fstream>
#include <iostream>

int main(int argc, char** argv) {
  namespace fs = std::filesystem;

  fs::path source_dir("./test_source_dir");
  fs::path dest_dir("./test_dest_dir");

  fs::remove_all(source_dir);
  fs::remove_all(dest_dir);
  fs::create_directory(source_dir);
  fs::create_directory(dest_dir);
  std::ofstream((source_dir/"1.txt").c_str()) << "source file 1";
  std::ofstream((source_dir/"2.txt").c_str()) << "source file 2";
  std::ofstream((dest_dir/"1.txt").c_str()) << "dest file 1";
  std::ofstream((dest_dir/"3.txt").c_str()) << "dest file 3";

  std::error_code error;
  fs::rename(source_dir, dest_dir, error);
  if (error) {
    std::cout << "error: " << error.message() << std::endl;
  } else {
    std::cout << "success" << std::endl;
  }

  return 0;
}
```

prints

```
error: Directory not empty
```

So do you want anything else changed or is this good as-is then?